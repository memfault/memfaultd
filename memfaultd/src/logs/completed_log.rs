//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use crate::mar::CompressionAlgorithm;
use std::path::PathBuf;
use uuid::Uuid;

/// CompletedLog represents a log that has been rotated and is ready to be moved into the MAR
/// staging area.
pub struct CompletedLog {
    pub path: PathBuf,
    pub cid: Uuid,
    pub next_cid: Uuid,
    pub compression: CompressionAlgorithm,
}
