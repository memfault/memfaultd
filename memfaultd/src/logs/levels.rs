//
// Copyright (c) Memfault, Inc.
// See License.txt for details
pub const LOG_LEVEL_EMERGENCY: &str = "EMERGENCY";
pub const LOG_LEVEL_ALERT: &str = "ALERT";
pub const LOG_LEVEL_CRITICAL: &str = "CRITICAL";
pub const LOG_LEVEL_ERROR: &str = "ERROR";
pub const LOG_LEVEL_WARN: &str = "WARN";
pub const LOG_LEVEL_NOTICE: &str = "NOTICE";
pub const LOG_LEVEL_INFO: &str = "INFO";
pub const LOG_LEVEL_DEBUG: &str = "DEBUG";

// systemd severity codes
pub const LOG_LEVEL_CODE_EMERGENCY: &str = "0";
pub const LOG_LEVEL_CODE_ALERT: &str = "1";
pub const LOG_LEVEL_CODE_CRITICAL: &str = "2";
pub const LOG_LEVEL_CODE_ERROR: &str = "3";
pub const LOG_LEVEL_CODE_WARN: &str = "4";
pub const LOG_LEVEL_CODE_NOTICE: &str = "5";
pub const LOG_LEVEL_CODE_INFO: &str = "6";
pub const LOG_LEVEL_CODE_DEBUG: &str = "7";
