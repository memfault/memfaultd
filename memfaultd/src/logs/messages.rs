//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use eyre::Result;
use ssf::Message;

use super::log_entry::LogEntry;

pub struct FlushLogsMsg;

impl Message for FlushLogsMsg {
    type Reply = Result<()>;
}

pub struct GetQueuedLogsMsg;

impl Message for GetQueuedLogsMsg {
    type Reply = Result<Vec<String>>;
}

pub struct RotateIfNeededMsg;

impl Message for RotateIfNeededMsg {
    type Reply = Result<bool>;
}

impl Message for LogEntry {
    type Reply = Result<()>;
}

pub struct RecoverLogsMsg;

impl Message for RecoverLogsMsg {
    type Reply = Result<bool>;
}

pub struct LogEntryMsg {
    pub entry: LogEntry,
    pub dropped_msg_count: usize,
}

impl LogEntryMsg {
    pub fn new(entry: LogEntry, dropped_msg_count: usize) -> Self {
        Self {
            entry,
            dropped_msg_count,
        }
    }
}

impl Message for LogEntryMsg {
    type Reply = Result<()>;
}
