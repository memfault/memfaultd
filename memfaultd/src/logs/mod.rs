//
// Copyright (c) Memfault, Inc.
// See License.txt for details
pub mod completed_log;
pub mod messages;
pub use completed_log::CompletedLog;
pub mod log_collector;
pub use log_collector::{LogCollector, LogCollectorConfig};
pub mod headroom;
pub use headroom::HeadroomLimiter;
#[cfg(feature = "systemd")]
pub mod journald_parser;
#[cfg(feature = "systemd")]
pub mod journald_provider;
pub mod levels;
pub mod log_entry;
mod log_file;
pub mod log_filter;
pub mod log_level_mapper;
pub mod log_to_metrics;
mod recovery;
#[cfg(feature = "syslog")]
pub mod syslog;
