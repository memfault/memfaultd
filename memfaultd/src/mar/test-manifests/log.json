{
    "schema_version": 1,
    "collection_time": {
        "timestamp": "1970-01-01T00:00:04.321Z",
        "uptime_ms": 800000,
        "elapsed_realtime_ms": 900000,
        "linux_boot_id": "230295cb-04d4-40b8-8624-ec37089b9b75",
        "boot_count": 0
    },
    "type": "linux-logs",
    "device": {
        "project_key": "INSERT_PROJECT_KEY_HERE",
        "hardware_version": "evt",
        "software_version": "1.0.0",
        "software_type": "linux-build",
        "device_serial": "DEMOSERIAL"
    },
    "metadata": {
        "format": { "id": "v1", "serialization": "json-lines" },
        "producer": { "id": "memfaultd", "version": "2.0.0" },
        "log_file_name": "logs.txt",
        "cid": { "uuid": "271e5249-1e3f-4c89-b4a5-ed96cc8f42a7" },
        "next_cid": { "uuid": "4276a26b-ba89-4c57-832f-d9e90e9629dd" }
    }
}
