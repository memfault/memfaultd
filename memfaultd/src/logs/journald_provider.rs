//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use crate::logs::journald_parser::{Journal, JournalRaw};

use eyre::Result;
use ssf::{MsgMailbox, Service, TaskService};

use std::{future::Future, pin::Pin};

use super::{log_collector::LogEntrySender, log_entry::LogEntry, messages::LogEntryMsg};

/// A log provider that reads log entries from Journald and sends them to a receiver.
pub struct JournaldLogProvider<J: JournalRaw> {
    journal: Journal<J>,
    extra_attr: Vec<String>,
    entry_sender: LogEntrySender,
}

impl<J: JournalRaw> JournaldLogProvider<J> {
    pub fn new(journal: J, entry_sender: MsgMailbox<LogEntryMsg>, extra_attr: Vec<String>) -> Self {
        let entry_sender = LogEntrySender::new(entry_sender);
        let journal = Journal::new(journal);

        Self {
            journal,
            extra_attr,
            entry_sender,
        }
    }

    pub async fn run_once(&mut self) -> Result<()> {
        self.journal.wait_for_entry().await?;
        for entry in self
            .journal
            .iter()
            .filter(|e| e.fields.contains_key("MESSAGE"))
        {
            // We would only fail here if 'MESSAGE' is not present. Which we verified above.
            let mut log_entry = LogEntry::try_from(entry)?;
            log_entry.filter_extra_fields(&self.extra_attr);

            self.entry_sender.send_entry(log_entry)?;
        }
        Ok(())
    }
}

impl<J: JournalRaw> Service for JournaldLogProvider<J> {
    fn name(&self) -> &str {
        "JournaldLogProvider"
    }
}

impl<J: JournalRaw> TaskService for JournaldLogProvider<J> {
    fn run_task(&mut self) -> Pin<Box<dyn Future<Output = std::result::Result<(), String>> + '_>> {
        Box::pin(async {
            self.run_once()
                .await
                .map_err(|e| format!("journald parser failed: {}", e))
        })
    }
}

#[cfg(test)]
mod test {
    use chrono::DateTime;
    use insta::{assert_json_snapshot, with_settings};
    use mockall::Sequence;
    use rstest::rstest;
    use ssf::ServiceMock;

    use super::*;

    use crate::logs::journald_parser::{JournalEntryRaw, MockJournalRaw};

    #[rstest]
    #[case("no_extra_attr".to_string(), vec![])]
    #[case("extra_attr".to_string(), vec!["EXTRA_FIELD".to_string()])]
    #[tokio::test]
    async fn test_happy_path(#[case] test_name: String, #[case] extra_attr: Vec<String>) {
        let mut journal_raw = MockJournalRaw::new();
        let mut seq = Sequence::new();

        let mut service = ServiceMock::new();

        journal_raw
            .expect_wait_for_entry()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Box::pin(async { Ok(()) }));
        journal_raw
            .expect_next_entry_available()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(true));
        journal_raw
            .expect_get_entry_field_data()
            .returning(|| Ok(Some(raw_journal_entry())));
        journal_raw
            .expect_next_entry_available()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(false));

        let mut provider = JournaldLogProvider::new(journal_raw, service.mbox.clone(), extra_attr);

        assert!(provider.run_once().await.is_ok());
        let entries = service.take_messages();
        assert_eq!(entries.len(), 1);
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(test_name, entries[0].entry);
        });
    }

    #[tokio::test]
    async fn test_channel_dropped() {
        let mut journal_raw = MockJournalRaw::new();
        let mut seq = Sequence::new();

        let service = ServiceMock::new();
        let mbox = service.mbox.clone();
        drop(service);

        journal_raw
            .expect_wait_for_entry()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Box::pin(async { Ok(()) }));
        journal_raw
            .expect_next_entry_available()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(true));
        journal_raw
            .expect_get_entry_field_data()
            .returning(|| Ok(Some(raw_journal_entry())));

        let mut provider = JournaldLogProvider::new(journal_raw, mbox, vec![]);

        assert!(provider.run_once().await.is_err());
    }

    fn raw_journal_entry() -> JournalEntryRaw {
        let fields = [
            "_SYSTEMD_UNIT=user@1000.service",
            "MESSAGE=audit: type=1400 audit(1713462571.968:7508): apparmor=\"DENIED\" operation=\"open\" class=\"file\" profile=\"snap.firefox.firefox\" name=\"/etc/fstab\" pid=10122 comm=\"firefox\" requested_mask=\"r\" denied_mask=\"r\" fsuid=1000 ouid=0",
            "EXTRA_FIELD=extra",
        ];

        let timestamp = DateTime::from_timestamp_millis(1337).unwrap();

        JournalEntryRaw::new(fields.iter().map(|s| s.to_string()).collect(), timestamp)
    }
}
