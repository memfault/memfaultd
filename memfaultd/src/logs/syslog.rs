//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Implements a simple syslog receiver UDP server than can be
//! used to forward messages from a syslog daemon like
//! syslogd or rsyslog to Memfault.
//!
//! Messages are expected to be either in the syslog RFC 5424 or
//! RFC 3164 format, with a single UDP packet per message.
//!
//! The maximum message size is 2048 bytes and longer messages
//! will be truncated.
//!
//! syslog message format RFCs:
//! RFC 3164: https://www.rfc-editor.org/rfc/rfc3164.html
//! RFC 5424: https://www.rfc-editor.org/rfc/rfc5424.html
//!
//! syslog over UDP RFC:
//! https://www.rfc-editor.org/rfc/rfc5426.html
//!
use std::collections::HashMap;
use std::net::{SocketAddr, UdpSocket};

use chrono::{Datelike, Utc};
use chrono_tz::Tz;

use eyre::{eyre, Result};
use log::warn;
use ssf::MsgMailbox;
use syslog_loose::{parse_message_with_year_exact_tz, ProcId, SyslogSeverity};

use crate::logs::{
    log_collector::LogEntrySender,
    log_entry::{LogData, LogEntry},
    messages::LogEntryMsg,
};
use crate::util::system::get_system_timezone;

use super::levels::{
    LOG_LEVEL_CODE_ALERT, LOG_LEVEL_CODE_CRITICAL, LOG_LEVEL_CODE_DEBUG, LOG_LEVEL_CODE_EMERGENCY,
    LOG_LEVEL_CODE_ERROR, LOG_LEVEL_CODE_INFO, LOG_LEVEL_CODE_NOTICE, LOG_LEVEL_CODE_WARN,
};

// From https://www.rfc-editor.org/rfc/rfc5426#section-3.2:
// "All syslog receivers SHOULD be able to receive datagrams
// with message sizes of up to and including 2048 octets."
const MAX_UDP_PACKET_SIZE: usize = 2048;

#[derive(Clone)]
pub struct SyslogServer {}

// TODO: Refactor to use TaskService
impl SyslogServer {
    pub fn run(bind_address: SocketAddr, sender: MsgMailbox<LogEntryMsg>) -> Result<()> {
        let sender = LogEntrySender::new(sender);
        let timezone = get_system_timezone().ok();
        let socket = UdpSocket::bind(bind_address)?;

        loop {
            // From https://www.rfc-editor.org/rfc/rfc5426#section-3.1:
            // "Each syslog UDP datagram MUST contain only one syslog message"
            let mut buf = [0; MAX_UDP_PACKET_SIZE];
            match socket.recv(&mut buf) {
                Ok(amt) => {
                    let message = String::from_utf8_lossy(&buf[..amt]);
                    if let Ok(log_entry) = Self::parse_syslog_message(&message, timezone) {
                        if sender.send_entry(log_entry).is_err() {
                            // An error indicates that the channel has been closed, we should
                            // kill this thread.
                            break;
                        }
                    }
                }
                Err(e) => warn!("Syslog server socket error: {}", e),
            }
        }

        Ok(())
    }

    fn priority_code_from_syslog_severity(syslog_severity: SyslogSeverity) -> String {
        match syslog_severity {
            SyslogSeverity::SEV_CRIT => LOG_LEVEL_CODE_CRITICAL.to_string(),
            SyslogSeverity::SEV_ALERT => LOG_LEVEL_CODE_ALERT.to_string(),
            SyslogSeverity::SEV_EMERG => LOG_LEVEL_CODE_EMERGENCY.to_string(),
            SyslogSeverity::SEV_ERR => LOG_LEVEL_CODE_ERROR.to_string(),
            SyslogSeverity::SEV_WARNING => LOG_LEVEL_CODE_WARN.to_string(),
            SyslogSeverity::SEV_INFO => LOG_LEVEL_CODE_INFO.to_string(),
            SyslogSeverity::SEV_DEBUG => LOG_LEVEL_CODE_DEBUG.to_string(),
            SyslogSeverity::SEV_NOTICE => LOG_LEVEL_CODE_NOTICE.to_string(),
        }
    }

    fn resolve_year((month, _date, _hour, _min, _sec): syslog_loose::IncompleteDate) -> i32 {
        let now = Utc::now();
        // If it is now January, messages from December are assumed to be from the previous year
        if now.month() == 1 && month == 12 {
            now.year() - 1
        } else {
            now.year()
        }
    }

    /// Parses a syslog message in either RFC 3164 or RFC 5424 format
    /// `system_timezone` is used as the timezone if it can be obtained
    /// from the system and there is not a timezone specified in the
    /// syslog timestamp
    fn parse_syslog_message(raw_message: &str, system_timezone: Option<Tz>) -> Result<LogEntry> {
        let syslog_entry = parse_message_with_year_exact_tz(
            raw_message,
            Self::resolve_year,
            system_timezone,
            syslog_loose::Variant::Either,
        )
        .map_err(|e| eyre!(e))?;

        let data = LogData {
            message: syslog_entry.msg.to_string(),
            pid: match syslog_entry.procid {
                Some(ProcId::PID(pid)) => Some(pid.to_string()),
                _ => None,
            },
            systemd_unit: match syslog_entry.appname {
                Some(appname) => Some(appname.to_string()),
                // Fallback to hostname if no appname is supplied
                // This is because busybox syslogd puts the application
                // name in the hostname position.
                None => syslog_entry.hostname.map(|hostname| hostname.to_string()),
            },
            priority: syslog_entry
                .severity
                .map(Self::priority_code_from_syslog_severity),
            original_priority: None,
            extra_fields: HashMap::new(),
        };

        let ts = syslog_entry
            .timestamp
            .ok_or_else(|| eyre!("Log has no timestamp"))?
            .into();

        Ok(LogEntry { ts, data })
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    use chrono_tz::Tz;
    use insta::assert_json_snapshot;
    use rstest::rstest;

    #[rstest]
    #[case(
        "<34>1 2023-05-15T14:30:45Z hostname systemd 1234 ID - [meta key=\"value\"] System started",
        Some(chrono_tz::UTC),
        "system_started_rfc_5424"
    )]
    #[case(
        "<34>May 15 14:30:45 hostname systemd[1234]: System started",
        Some(chrono_tz::America::Los_Angeles),
        "system_started_rfc_3164"
    )]
    #[case(
        "<13>May 15 14:30:46 hostname kernel: Kernel panic detected!!",
        Some(chrono_tz::America::Los_Angeles),
        "kernel_panic_rfc_3164"
    )]
    fn test_read_syslog_message(
        #[case] input: &str,
        #[case] timezone: Option<Tz>,
        #[case] snapshot_name: &str,
    ) {
        let result = SyslogServer::parse_syslog_message(input, timezone).unwrap();

        assert_json_snapshot!(snapshot_name, result);
    }

    #[test]
    fn test_missing_timestamp() {
        let input = "<34>hostname systemd[1234]: System started";

        let result = SyslogServer::parse_syslog_message(input, Some(chrono_tz::UTC));
        assert!(result.is_err());
    }
}
