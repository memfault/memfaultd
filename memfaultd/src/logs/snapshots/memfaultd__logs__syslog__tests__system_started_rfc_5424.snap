---
source: memfaultd/src/logs/syslog.rs
expression: result
---
{
  "ts": "2023-05-15T14:30:45+00:00",
  "data": {
    "MESSAGE": "[meta key=\"value\"] System started",
    "_PID": "1234",
    "_SYSTEMD_UNIT": "systemd",
    "PRIORITY": "2"
  }
}
